File: totoggle_go/internal/toggle/rule.go
package toggle

import (
	"bytes"
	"encoding/json"
	"strings"
)

// RuleType is one of the 7 activation-rule types the server supports
// (server/internal/app/domain/entity/activation_rule.go#ActivationRuleType) — a typed string so
// a typo can't silently compile as a valid, never-matching type the way a bare string could.
type RuleType string

const (
	RuleTypePercentage RuleType = "percentage"
	RuleTypeAttribute  RuleType = "attribute"
	RuleTypeUserID     RuleType = "user_id"
	RuleTypeIP         RuleType = "ip"
	RuleTypeCountry    RuleType = "country"
	RuleTypeTime       RuleType = "time"
	RuleTypeCohort     RuleType = "cohort"
)

// ActivationRule is a value object: Type and Value together define a condition, and neither is
// meaningful alone (a Value with no Type, or vice versa, is never a valid rule).
type ActivationRule struct {
	Type   RuleType        `json:"type"`
	Value  string          `json:"value"`
	Config json.RawMessage `json:"config"`
}

// ContextKey returns a canonical configured provider key. A catalogue is untrusted input even
// after the server has validated it, so incompatible type/key pairs fail closed in the SDK before
// a resolver is invoked.
func (r ActivationRule) ContextKey() (string, bool) {
	if r.Type == RuleTypeTime {
		return "", len(r.Config) == 0 || bytes.Equal(bytes.TrimSpace(r.Config), []byte("null"))
	}
	var config struct {
		ContextKey string `json:"context_key"`
	}
	if json.Unmarshal(r.Config, &config) != nil || !validContextKey(r.Type, config.ContextKey) {
		return "", false
	}
	return config.ContextKey, true
}

func validContextKey(ruleType RuleType, key string) bool {
	if strings.HasPrefix(key, "attributes.") {
		return len(strings.TrimPrefix(key, "attributes.")) > 0 &&
			(ruleType == RuleTypePercentage || ruleType == RuleTypeAttribute)
	}
	return (ruleType == RuleTypePercentage && key == "rollout_key") ||
		(ruleType == RuleTypeUserID && key == "user_id") ||
		(ruleType == RuleTypeIP && key == "ip") ||
		(ruleType == RuleTypeCountry && key == "country") ||
		(ruleType == RuleTypeCohort && key == "cohort")
}

// IsEmpty reports whether this is "no rule configured" (both fields blank).
func (r ActivationRule) IsEmpty() bool {
	return r.Type == "" && r.Value == ""
}

// IsValid reports whether this rule has both a type and a value — required for it to be
// evaluated at all.
func (r ActivationRule) IsValid() bool {
	return r.Type != "" && r.Value != ""
}

File: totoggle_node/src/internal/toggle/rule.ts
/** The 7 activation-rule types the server supports. A string-literal union, not an enum — lets
 * a parsed `type` field compare structurally without an extra mapping step. */
export const RULE_TYPES = [
  "percentage",
  "attribute",
  "user_id",
  "ip",
  "country",
  "time",
  "cohort",
] as const;

export type RuleType = (typeof RULE_TYPES)[number];

/** ActivationRule is a value object: type and value together define a condition, and neither is
 * meaningful alone (a value with no type, or vice versa, is never a valid rule). A plain
 * interface, not a class — it doubles as the fetch DTO shape with no separate mapping layer. */
export interface ActivationRule {
  readonly type: string;
  readonly value: string;
  readonly config?: { readonly context_key?: string } | null;
}

/**
 * Validates the type/context-key pair before a resolver is touched. Catalogue data crosses a
 * trust boundary, so SDK evaluation repeats this small canonical contract and fails closed when
 * an invalid server payload somehow reaches the cache.
 */
export function hasCanonicalContextKey(rule: ActivationRule): boolean {
  if (rule.type === "time") return rule.config === undefined || rule.config === null;

  const key = rule.config?.context_key;
  if (key === undefined || key === "") return false;
  if (key.startsWith("attributes.")) {
    return key.length > "attributes.".length && (rule.type === "percentage" || rule.type === "attribute");
  }
  return (rule.type === "percentage" && key === "rollout_key")
    || (rule.type === "user_id" && key === "user_id")
    || (rule.type === "ip" && key === "ip")
    || (rule.type === "country" && key === "country")
    || (rule.type === "cohort" && key === "cohort");
}

/** Reports whether this is "no rule configured" (both fields blank). */
export function isEmpty(rule: ActivationRule): boolean {
  return rule.type === "" && rule.value === "";
}

/** Reports whether this rule has both a type and a value — required for it to be evaluated at
 * all. */
export function isValid(rule: ActivationRule): boolean {
  return rule.type !== "" && rule.value !== "";
}
